File: app/atem_controller.py
# app/atem_controller.py
import PyATEMMax


class AtemController:
    def __init__(self):
        self.switcher = PyATEMMax.ATEMMax()
        self.connected = False
        self.current_input = None
        self.ip = None
        self._started = False  # True once switcher.connect() has spawned its comms threads
        self.switcher.registerEvent(self.switcher.atem.events.connect, self._on_connect)
        self.switcher.registerEvent(self.switcher.atem.events.disconnect, self._on_disconnect)

    def _on_connect(self, params):
        self.connected = True

    def _on_disconnect(self, params):
        self.connected = False
        self.current_input = None

    def connect(self, ip):
        # PyATEMMax.connect() spawns comms + event threads and reconnects internally
        # forever — it's meant to be called ONCE. Calling it again (e.g. the user changes
        # the IP) without tearing down the previous attempt stacks threads that all share
        # one UDP socket, racing into OSError EOPNOTSUPP. So always disconnect first.
        self.ip = ip
        if self._started:
            try:
                self.switcher.disconnect()
            except Exception:
                pass
            self._started = False
        self.switcher.connect(ip)
        self._started = True

    def disconnect(self):
        if self._started:
            try:
                self.switcher.disconnect()
            except Exception:
                pass
            self._started = False
        self.connected = False
        self.current_input = None

    def cut_to(self, atem_input, me_index=0):
        if not self.connected:
            return
        if self.current_input == atem_input:
            return
        self.switcher.setPreviewInputVideoSource(me_index, atem_input)
        self.switcher.execCutME(me_index)
        self.current_input = atem_input

    def auto_to(self, atem_input, me_index=0):
        if not self.connected:
            return
        if self.current_input == atem_input:
            return
        self.switcher.setPreviewInputVideoSource(me_index, atem_input)
        self.switcher.execAutoME(me_index)
        self.current_input = atem_input

    def get_status(self):
        return {"connected": self.connected, "ip": self.ip, "currentInput": self.current_input}

    # NOTE: no reconnect loop here on purpose. PyATEMMax's comms thread keeps the
    # connection alive and retries forever after a single connect(), so an external
    # reconnect loop only stacks duplicate comms threads (the EOPNOTSUPP bug).

File: reference/mic-cam-switcher-node/src/atemController.js
const { Atem } = require('atem-connection');

class AtemController {
  constructor() {
    this.atem = new Atem();
    this.connected = false;
    this.currentInput = null;
    this.ip = null;

    this.atem.on('connected', () => {
      this.connected = true;
      console.log(`[ATEM] connected to ${this.ip}`);
    });
    this.atem.on('disconnected', () => {
      this.connected = false;
      console.log('[ATEM] disconnected');
    });
    this.atem.on('error', (e) => {
      console.error('[ATEM] error', e);
    });
  }

  connect(ip) {
    this.ip = ip;
    try {
      this.atem.connect(ip);
    } catch (e) {
      console.error('[ATEM] connect failed', e);
    }
  }

  disconnect() {
    try {
      this.atem.disconnect();
    } catch (e) {
      /* ignore */
    }
  }

  // meIndex is the mix effect bus, 0 = Program bus 1 (most single-room setups use ME 0)
  async cutTo(atemInput, meIndex = 0) {
    if (!this.connected) return;
    if (this.currentInput === atemInput) return;
    try {
      await this.atem.setPreviewInput(atemInput, meIndex);
      await this.atem.cut(meIndex);
      this.currentInput = atemInput;
    } catch (e) {
      console.error('[ATEM] cut failed', e);
    }
  }

  async autoTo(atemInput, meIndex = 0) {
    if (!this.connected) return;
    if (this.currentInput === atemInput) return;
    try {
      await this.atem.setPreviewInput(atemInput, meIndex);
      await this.atem.autoTransition(meIndex);
      this.currentInput = atemInput;
    } catch (e) {
      console.error('[ATEM] auto transition failed', e);
    }
  }

  getStatus() {
    return {
      connected: this.connected,
      ip: this.ip,
      currentInput: this.currentInput,
    };
  }
}

module.exports = AtemController;
